scripts/build.py

```diff
@@ -715,63 +908,76 @@ def generate_build_directory():
-        print 'Creating new %s folder' % TEMPLATES_STAGING_DIR
-        build_files(
-            TEMPLATES_DEV_DIR_CORE, TEMPLATES_STAGING_DIR, hashes)
-    else:
-        # If there is an existing staging dir, rebuild all HTML files.
-        print 'Staging directory exists, re-building all HTML files'
-        build_files(
-            TEMPLATES_DEV_DIR_CORE, TEMPLATES_STAGING_DIR, hashes,
-            file_formats=('.html',))
-        recently_changed_filenames = get_recently_changed_filenames(
-            TEMPLATES_DEV_DIR_CORE, TEMPLATES_OUT_DIR)
-        if recently_changed_filenames:
-            # Only re-build files that have changed since last build.
-            print ("Re-building recently changed files at %s"
-                   % TEMPLATES_DEV_DIR_CORE)
-            rebuild_new_files(
-                TEMPLATES_DEV_DIR_CORE, TEMPLATES_STAGING_DIR,
-                recently_changed_filenames, hashes)
-            # Clean up files in staging dir that have been removed in DEV dir.
-            remove_deleted_files(TEMPLATES_DEV_DIR_CORE, TEMPLATES_STAGING_DIR)
-        else:
-            print 'No changes detected. Using previously built files.'
-    # Copy built files from staging dir to build dir.
-    copy_files_source_to_target(
-        TEMPLATES_STAGING_DIR, TEMPLATES_OUT_DIR, hashes)
-
+    ensure_directory_exists(HASHES_JS_FILEPATH)
+    with open(HASHES_JS_FILEPATH, 'w+') as hashes_js_file:
+        write_to_file_stream(
+            hashes_js_file, get_hashes_json_file_contents(hashes))
+    # Update hash dict with newly created hashes.js.
+    hashes.update({HASHES_JS_FILENAME: generate_md5_hash(HASHES_JS_FILEPATH)})
+    # Make sure /assets/hashes.js is available to the frontend.
+    _ensure_files_exist([HASHES_JS_FILEPATH])
+
+    # Build files in /extensions and copy them into staging directory.
+    build_tasks += generate_build_task_to_build_directory(
+        EXTENSIONS_DIRNAMES_TO_DIRPATHS, hashes)
+    # Minify all template files and copy them into build/templates/head.
+    build_tasks += generate_build_task_to_build_directory(
+        TEMPLATES_CORE_DIRNAMES_TO_DIRPATHS, hashes)
+    _execute_tasks(build_tasks)
+
+    # Copy all files from staging directory to production directory (/build).
+    COPY_INPUT_DIRS = [
+        ASSETS_DEV_DIR, EXTENSIONS_DIRNAMES_TO_DIRPATHS['staging_dir'],
+        TEMPLATES_CORE_DIRNAMES_TO_DIRPATHS['staging_dir'],
+        THIRD_PARTY_GENERATED_DEV_DIR]
+    COPY_OUTPUT_DIRS = [
+        ASSETS_OUT_DIR, EXTENSIONS_DIRNAMES_TO_DIRPATHS['out_dir'],
+        TEMPLATES_CORE_DIRNAMES_TO_DIRPATHS['out_dir'],
+        THIRD_PARTY_GENERATED_OUT_DIR]
+    assert len(COPY_INPUT_DIRS) == len(COPY_OUTPUT_DIRS)
+    for i in xrange(len(COPY_INPUT_DIRS)):
+        copy_tasks += generate_copy_tasks_to_copy_from_source_to_target(
+            COPY_INPUT_DIRS[i], COPY_OUTPUT_DIRS[i], hashes)
+    _execute_tasks(copy_tasks)
+
+    for i in xrange(len(COPY_INPUT_DIRS)):
```

Comment:
Actually -- this function is quite long already. Can we extract the verification stuff into a separate private function called _verify_build()?

Reply:
Moved verification processes to its own private function!